firmware/library/L1_Drivers/adc.hpp

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <cstdint>
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/pin_configure.hpp"
+#include "L2_Utilities/macros.hpp"
+
+class AdcInterface
+{
+ public:
+    virtual void Initialize(uint32_t adc_clk_hz = 1'000'000) = 0;
+    virtual void Conversion() = 0;
+    virtual uint16_t ReadResult() = 0;
+    virtual bool FinishedConversion() = 0;
+};
+
+class Adc : public AdcInterface
+{
+ public:
+    enum class Channel : uint8_t
+    {
+        kChannel0 = 0,
+        kChannel1,
+        kChannel2,
+        kChannel3,
+        kChannel4,
+        kChannel5,
+        kChannel6,
+        kChannel7,
+        kMaxNumOfPins
+    };
+    enum AdcPortsMap : uint8_t
+    {
+        kChannel0Port = 0,
+        kChannel1Port = 0,
+        kChannel2Port = 0,
+        kChannel3Port = 0,
+        kChannel4Port = 1,
+        kChannel5Port = 1,
+        kChannel6Port = 0,
+        kChannel7Port = 0
+    };
+    enum AdcPinsMap : uint8_t
+    {
+        kChannel0Pin = 23,
+        kChannel1Pin = 24,
+        kChannel2Pin = 25,
+        kChannel3Pin = 26,
+        kChannel4Pin = 30,
+        kChannel5Pin = 31,
+        kChannel6Pin = 12,
+        kChannel7Pin = 13
+    };
+    enum ControlBit : uint8_t
+    {
+        kBurstMode = 16,
+        kPowerUp = 21,
+        kStart = 24
+    };
+
+    static constexpr uint8_t kMaxNumOfPins =
+        static_cast<uint8_t>(Channel::kMaxNumOfPins);
+    const uint8_t kAdcPorts[kMaxNumOfPins] = {
+        kChannel0Port, kChannel1Port, kChannel2Port, kChannel3Port,
+        kChannel4Port, kChannel5Port, kChannel6Port, kChannel7Port
+    };
+    const uint8_t kAdcPins[kMaxNumOfPins] = {
+        kChannel0Pin, kChannel1Pin, kChannel2Pin, kChannel3Pin,
+        kChannel4Pin, kChannel5Pin, kChannel6Pin, kChannel7Pin
+    };
+
+    static LPC_ADC_TypeDef *adc_base;
+    static void BurstMode(bool burst_mode_is_on = false)
+    {
+        (burst_mode_is_on = false) ?
+            (adc_base->CR &= ~(1 << ControlBit::kBurstMode)) :
+            (adc_base->CR |= (1 << ControlBit::kBurstMode));
+    }
+    explicit constexpr Adc(Adc::Channel channel_bit)
+        : adc_(adc_pin_),
+          adc_pin_(PinConfigure(kAdcPorts[static_cast<uint8_t>(channel_bit)],
+                                kAdcPins[static_cast<uint8_t>(channel_bit)]))
+    {
+        adc_base->CR |= (1 << (static_cast<uint8_t>(channel_bit)));
```

Comment:
Do not manipulate hardware in constructors. You can do this in your init method.